cmd/registry/scoring/parser_test.go

```diff
@@ -0,0 +1,911 @@
+	tests := []struct {
+		desc       string
+		minValue   int32
+		maxValue   int32
+		thresholds []*rpc.NumberThreshold
+	}{
+		{
+			desc:     "percentage thresholds",
+			minValue: 0,
+			maxValue: 100,
+			thresholds: []*rpc.NumberThreshold{
+				{
+					Severity: rpc.Severity_ALERT,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 0,
+						Max: 30,
+					},
+				},
+				{
+					Severity: rpc.Severity_WARNING,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 31,
+						Max: 60,
+					},
+				},
+				{
+					Severity: rpc.Severity_OK,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 61,
+						Max: 100,
+					},
+				},
+			},
+		},
+		{
+			desc:     "integer thresholds",
+			minValue: -50,
+			maxValue: 50,
+			thresholds: []*rpc.NumberThreshold{
+				{
+					Severity: rpc.Severity_ALERT,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: -50,
+						Max: -20,
+					},
+				},
+				{
+					Severity: rpc.Severity_WARNING,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: -19,
+						Max: 10,
+					},
+				},
+				{
+					Severity: rpc.Severity_OK,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 11,
+						Max: 50,
+					},
+				},
+			},
+		},
+		{
+			desc:     "descending score thresholds",
+			minValue: 0,
+			maxValue: 100,
+			thresholds: []*rpc.NumberThreshold{
+				{
+					Severity: rpc.Severity_ALERT,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 61,
+						Max: 100,
+					},
+				},
+				{
+					Severity: rpc.Severity_WARNING,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 31,
+						Max: 60,
+					},
+				},
+				{
+					Severity: rpc.Severity_OK,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 0,
+						Max: 30,
+					},
+				},
+			},
+		},
+		{
+			desc:     "distributed severity",
+			minValue: 0,
+			maxValue: 100,
+			thresholds: []*rpc.NumberThreshold{
+				{
+					Severity: rpc.Severity_ALERT,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 0,
+						Max: 10,
+					},
+				},
+				{
+					Severity: rpc.Severity_ALERT,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 91,
+						Max: 100,
+					},
+				},
+				{
+					Severity: rpc.Severity_WARNING,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 11,
+						Max: 30,
+					},
+				},
+				{
+					Severity: rpc.Severity_WARNING,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 71,
+						Max: 90,
+					},
+				},
+				{
+					Severity: rpc.Severity_OK,
+					Range: &rpc.NumberThreshold_NumberRange{
+						Min: 31,
+						Max: 70,
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			gotErrs := validateNumberThresholds(test.thresholds, test.minValue, test.maxValue)
+			if len(gotErrs) > 0 {
+				t.Errorf("validateNumberThresholds() returned unexpected errors: %s", gotErrs)
+			}
+		})
+	}
+
+}
+
+func TestValidateNumberThresholdsError(t *testing.T) {
+	tests := []struct {
```

Comment:
Added a couple more conditions:
- Check that all the oneOf fields are set (`formula` and `type`).
- Check that `min_value` < `max_value` in `IntegerType` definition
- Check that `min` <= `max` in `NumberThreshold.Range`

For the two cases that you mentioned:
- `a threshold has minValue > maxValue` -> the added check should take care of this.
-  `the artifact has incomplete information like unspecified/nil number ranges in a threshold` -> this should be ideally taken care of in the proto. We can make the range field required (currently it's not):
```
message NumberThreshold {
  // Represents the severity level of the value of the score.
  Severity severity = 1 [(google.api.field_behavior) = REQUIRED];

  message NumberRange {
    int32 min = 1 [(google.api.field_behavior) = REQUIRED];
    int32 max = 2 [(google.api.field_behavior) = REQUIRED];
  }

  // The boundaries specified in this field will be included in the mentioned
  // severity level. For examples see: PercentType, IntegerType.
  NumberRange range = 2; [(google.api.field_behavior) = REQUIRED];
}
```

This will be inline with how the BooleanThreshold is designed:
```
message BooleanThreshold {
  // Represents the severity level of the value of the score.
  Severity severity = 1 [(google.api.field_behavior) = REQUIRED];

  // The value field will be used to determine the severity
  // level of the generated score.
  bool value = 2 [(google.api.field_behavior) = REQUIRED];
}
```

Reply:
Do field annotations enforce anything? I was under the impression that we need to validate manually to make sure our annotations are being followed, i.e. they're documentation. I'm not 100% sure though.

Reply:
You are right, the field annotations don't enforce anything. (I just tried to upload a ScoreDefinition which violated a required field).

I thought that they served the same purpose as writing `required string name` in proto2, but it's not true. We will have to manually check the existence of required fields.

Reply:
`the artifact has incomplete information like unspecified/nil number ranges in a threshold` -> in this case, `NumberThreshold.GetRange().GetMin()` will return default value `0` and it will be similar to the case where `min` and `max` in a range are not set. 

I added a test case for this particular case: "missing range"